### PR TITLE
[Merged by Bors] - Document `PipelineCache` and related types

### PR DESCRIPTION
**Pull request by djeedai**
# Objective

Document `PipelineCache` and a few other related types.

## Solution

Add documenting comments to `PipelineCache` and a few other related
types in the same file.

### CODE REVIEW

File: crates/bevy_render/src/render_resource/pipeline_cache.rs [hunk 1]
```diff
@@ -25,6 +25,9 @@ enum PipelineDescriptor {
     ComputePipelineDescriptor(Box<ComputePipelineDescriptor>),
 }
 
+/// Union of a render and compute pipeline.
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/render_resource/pipeline_cache.rs:28]:
```suggestion
/// A pipeline used to send data to the GPU.
```

The original was redundant, and `union` is a critically distinct keyword in Rust.

File: crates/bevy_render/src/render_resource/pipeline_cache.rs [hunk 2]
```diff
@@ -72,7 +89,7 @@ impl CachedPipelineState {
 }
 
 #[derive(Default)]
-pub struct ShaderData {
+struct ShaderData {
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/render_resource/pipeline_cache.rs:92]: This change shouldn't be part of a docs PR.

**Comment by alice-i-cecile** [comment]
Two small concerns, then this LGTM.

File: crates/bevy_render/src/render_resource/pipeline_cache.rs (hunk 2 above)
Comment (djeedai) [reply crates/bevy_render/src/render_resource/pipeline_cache.rs]: Reverted.

**Comment by djeedai** [comment]
@alice-i-cecile for final (?) review after all fixed, thanks!

File: crates/bevy_render/src/render_resource/pipeline_cache.rs (hunk 2 above)
Comment (djeedai) [reply crates/bevy_render/src/render_resource/pipeline_cache.rs]: Actually, opened a separate PR for this #5609

**Comment by djeedai** [comment]
Tiny rant : changes like #3979 that introduced a large part of this code should not have been allowed to merge without a _single_ comment on the entire file. I don't know if there's a policy now, but if there's none I'd like to suggest raising the bar a bit and strongly suggesting (if not requiring) at least a minimal documenting of new features.

**Review by alice-i-cecile — APPROVED** [review]
@bevyengine/rendering-team can I get a review from one of y'all? I'm happy with this from a docs perspective but not fully confident about the domain specific content.

File: crates/bevy_render/src/render_resource/pipeline_cache.rs [hunk 3]
```diff
@@ -25,6 +25,9 @@ enum PipelineDescriptor {
     ComputePipelineDescriptor(Box<ComputePipelineDescriptor>),
 }
 
+/// A pipeline used to send data to the GPU.
```
Comment (cart) [inline crates/bevy_render/src/render_resource/pipeline_cache.rs:28]: This is a mischaracterization. Pipelines aren't really about sending data to the GPU, they are about defining the data layout and shader logic for a specific GPU task. See https://docs.rs/wgpu/latest/wgpu/struct.RenderPipeline.html for details.

**Review by cart — COMMENTED** [review]
Generally looks good. Just one comment

File: crates/bevy_render/src/render_resource/pipeline_cache.rs (hunk 3 above)
Comment (alice-i-cecile) [reply crates/bevy_render/src/render_resource/pipeline_cache.rs:28]: Ah, thanks! This is my fault, and I was hoping someone with more rendering expertise would double check.

Comment (djeedai) [reply crates/bevy_render/src/render_resource/pipeline_cache.rs]: Fixed. Sorry I had forgotten about this PR.

**Pull request closed without merging**